src/NetAnalyzers/Core/Microsoft.NetCore.Analyzers/Performance/UseStringContainsCharOverloadWithSingleCharacters.Fixer.cs

```diff
@@ -0,0 +1,67 @@
+﻿// Copyright (c) Microsoft.  All Rights Reserved.  Licensed under the Apache License, Version 2.0.  See License.txt in the project root for license information.
+
+using System.Collections.Immutable;
+using System.Composition;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.CodeAnalysis;
+using Microsoft.CodeAnalysis.CodeActions;
+using Microsoft.CodeAnalysis.CodeFixes;
+using Microsoft.CodeAnalysis.Editing;
+using Microsoft.CodeAnalysis.Operations;
+
+namespace Microsoft.NetCore.Analyzers.Performance
+{
+    [ExportCodeFixProvider(LanguageNames.CSharp, LanguageNames.VisualBasic), Shared]
+    public sealed class UseStringContainsCharOverloadWithSingleCharactersCodeFix : CodeFixProvider
+    {
+        public override ImmutableArray<string> FixableDiagnosticIds => ImmutableArray.Create(
+            UseStringContainsCharOverloadWithSingleCharactersAnalyzer.CA1839);
+
+        public override async Task RegisterCodeFixesAsync(CodeFixContext context)
+        {
+            var root = await context.Document.GetSyntaxRootAsync(context.CancellationToken).ConfigureAwait(false);
+            var violatingNode = root.FindNode(context.Span, getInnermostNodeForTie: true);
+
+            context.RegisterCodeFix(new ReplaceStringLiteralWithCharLiteralCodeAction(context.Document, violatingNode), context.Diagnostics);
+        }
+
+        public override FixAllProvider GetFixAllProvider()
+        {
+            return WellKnownFixAllProviders.BatchFixer;
+        }
+
+        private class ReplaceStringLiteralWithCharLiteralCodeAction : CodeAction
+        {
+            private readonly Document _document;
+            private readonly SyntaxNode _sourceStringLiteral;
+
+            public override string Title => MicrosoftNetCoreAnalyzersResources.ReplaceStringLiteralWithCharLiteralCodeActionTitle;
+
+            public override string EquivalenceKey => nameof(ReplaceStringLiteralWithCharLiteralCodeAction);
+
+            public ReplaceStringLiteralWithCharLiteralCodeAction(Document document, SyntaxNode sourceStringLiteral)
+            {
+                _document = document;
+                _sourceStringLiteral = sourceStringLiteral;
+            }
+
+            protected override async Task<Document> GetChangedDocumentAsync(CancellationToken cancellationToken)
+            {
+                var editor = await DocumentEditor.CreateAsync(_document, cancellationToken).ConfigureAwait(false);
+                var semanticModel = await _document.GetSemanticModelAsync(cancellationToken).ConfigureAwait(false);
+                if (semanticModel.GetOperation(_sourceStringLiteral, cancellationToken) is ILiteralOperation literalOperation && literalOperation.ConstantValue.HasValue)
+                {
+                    if (literalOperation.ConstantValue.Value is string stringValue && char.TryParse(stringValue, out char value))
```

Comment:
Haha i'm absolutely not following here - Are you saying i could somehow avoid loading the semantic model by passing the fixed literal node together with analyzer?

Anyhow... can you go into detail - maybe link another CodeFix where this happens?

Reply:
@MeikTranel In the analyzer, when you report a diagnostic, you can pass a dictionary which can retrieved in the codefix. See the `properties` parameter in `context.ReportDiagnostic`.

In the fixer, you can access it through `Properties` property of the diagnostic.

It's like you're caching things you already did in the analyzer to not do it again in the fixer.

Reply:
Haha nice. That feels a little sketchy, but i'll give it a try. Itll probably be another day though since the rebase turned out to be a little more sketchy than expected (autogenerated files are fun :smile: )

Reply:
@Youssef1313 i think i've got though it feels sketchy NGL